src/NHibernate/Engine/ActionQueue.cs

```diff
@@ -505,106 +507,172 @@ private class InsertActionSorter
+					{
+						var prevBatchIdentifier = _orderedBatches[j];
+						if (prevBatchIdentifier.ParentEntityNames.Contains(entityName))
+						{
+							_orderedBatches.RemoveAt(i);
+							_orderedBatches.Insert(j, batchIdentifier);
+						}
+					}
+
+					//Make sure that parent entries are not after children
+					for (var j = i + 1; j < _orderedBatches.Count; j++)
 					{
-						_actionQueue.insertions.Add(action);
+						var nextBatchIdentifier = _orderedBatches[j];
+						//Take care of unidirectional @OneToOne associations but exclude bidirectional @ManyToMany
+						if (nextBatchIdentifier.ChildEntityNames.Contains(entityName) &&
+							!batchIdentifier.ChildEntityNames.Contains(nextBatchIdentifier.EntityName))
+						{
+							_orderedBatches.RemoveAt(i);
+							_orderedBatches.Insert(j, batchIdentifier);
+						}
 					}
 				}
-			}
 
-			private int GetBatchNumber(EntityInsertAction action, string entityName)
-			{
-				int batchNumber;
-				if (_latestBatches.TryGetValue(entityName, out batchNumber))
+				// now rebuild the insertions list. There is a batch for each entry in the batches list.
+				foreach (var batchIdentifier in _orderedBatches)
 				{
-					// There is already an existing batch for this type of entity.
-					// Check to see if the latest batch is acceptable.
-					if (IsProcessedAfterAllAssociatedEntities(action, batchNumber))
-						return batchNumber;
+					var batch = _actionBatches[batchIdentifier];
+					_actionQueue.insertions.AddRange(batch);
 				}
-				
-				// add an entry for this type of entity.
-				// we can be assured that all referenced entities have already
-				// been processed,
-				// so specify that this entity is with the latest batch.
-				// doing the batch number before adding the name to the list is
-				// a faster way to get an accurate number.
-
-				batchNumber = _actionBatches.Count;
-				_latestBatches[entityName] = batchNumber;
-				return batchNumber;
 			}
 
-			private bool IsProcessedAfterAllAssociatedEntities(EntityInsertAction action, int latestBatchNumberForType)
+			/// <summary>
+			/// Add parent and child entity names so that we know how to rearrange dependencies
+			/// </summary>
+			/// <param name="action">The action being sorted</param>
+			/// <param name="batchIdentifier">The batch identifier of the entity affected by the action</param>
+			private void AddParentChildEntityNames(EntityInsertAction action, BatchIdentifier batchIdentifier)
 			{
 				var propertyValues = action.State;
-				var propertyTypes = action.Persister.ClassMetadata.PropertyTypes;
+				// NH: it seems we support custom persister which does not provide meta-data (see CustomPersister in NHibernate.DomainModel).
+				var propertyTypes = action.Persister.ClassMetadata?.PropertyTypes;
+				if (propertyTypes == null)
+				{
+					log.WarnFormat("Entity {0} persister does not provide meta-data, its batch ordering may be wrong and cause a failure.",
+						batchIdentifier.EntityName);
+					return;
```

Comment:
They have "solved" it the same way: no ordering in case the persister does not provide class meta data. But I have noticed that on their side, their custom provider does provide an entity meta model while ours does not. Maybe syncing our custom provider with their and fallback-ing to the entity meta model could work.